### PR TITLE
[SPARK-48016][SQL] Binary Arithmetic operators should include the evalMode when makeCopy

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

 Currently, the following query will throw DIVIDE_BY_ZERO error instead of returning null
 ```
SELECT try_divide(1, decimal(0)); 
```

This is caused by the rule `DecimalPrecision`:
```
case b @ BinaryOperator(left, right) if left.dataType != right.dataType =>
  (left, right) match {
 ...
    case (l: Literal, r) if r.dataType.isInstanceOf[DecimalType] &&
        l.dataType.isInstanceOf[IntegralType] &&
        literalPickMinimumPrecision =>
      b.makeCopy(Array(Cast(l, DataTypeUtils.fromLiteral(l)), r)) 
```
The result of the above makeCopy will contain `ANSI` as the `evalMode`, instead of `TRY`.
This PR is to fix this bug.

### Why are the changes needed?

Bug fix in try_* functions.

### Does this PR introduce _any_ user-facing change?

Yes, it fixes a long-standing bug in the try_divide function. 

### How was this patch tested?

New UTs
### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala [hunk 1]
```diff
@@ -190,7 +190,7 @@ case class Abs(child: Expression, failOnError: Boolean = SQLConf.get.ansiEnabled
 abstract class BinaryArithmetic extends BinaryOperator
   with NullIntolerant with SupportQueryContext {
 
-  protected val evalMode: EvalMode.Value
+  protected[sql] val evalMode: EvalMode.Value
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala:193]: This is to make sure we can access `evalMode` in tests.

**Review by dongjoon-hyun — COMMENTED** [review]
Could you fix the UT failure, @gengliangwang ?
```
[info] *** 1 TEST FAILED ***
[error] Failed: Total 7392, Failed 1, Errors 0, Passed 7391, Ignored 5
[error] Failed tests:
[error] 	org.apache.spark.sql.catalyst.expressions.CanonicalizeSuite
[error] (catalyst / Test / test) sbt.TestsFailedException: Tests unsuccessful
```

**Review by cloud-fan — APPROVED** [review]
good catch!

**Review by HyukjinKwon — APPROVED** [review]
LGTM pending CI

**Comment by gengliangwang** [comment]
I am starting a new approach in https://github.com/apache/spark/pull/46286.
Fixing the `makeCopy` can be tricky. The test failure in CanonicalizeSuite seems a bit complicated to fix.

**Pull request closed without merging**